File: sdks/python/src/verdocs/resources/invitations.py
"""Organization invitation operations (js-sdk: Organizations/Invitations.ts).

Invitations are keyed by email within an organization and are one-time-use:
once accepted, declined, or deleted, the token stops working. get(), accept(),
and decline() authenticate with the invite token rather than a session, so
they work on an endpoint with no token set.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..models.core import OrganizationInvitation
from ..models.organizations import InvitationAcceptParams, InvitationCreateParams, InvitationUpdateParams
from ..models.users import AuthenticateResponse

if TYPE_CHECKING:
    from .._endpoint import AsyncVerdocsEndpoint, VerdocsEndpoint

_INVITATIONS_PATH = "/v2/organization-invitations"


def _write_body(params: InvitationCreateParams | InvitationUpdateParams | InvitationAcceptParams) -> dict[str, Any]:
    return params.model_dump(mode="json", exclude_unset=True)


class Invitations:
    """Organization invitation calls for a sync endpoint."""

    def __init__(self, endpoint: VerdocsEndpoint) -> None:
        self._endpoint = endpoint

    def list(self) -> list[OrganizationInvitation]:
        """Get the pending invitations for the caller's organization via GET /v2/organization-invitations.

        The caller must be an admin or owner. Mirrors js-sdk getOrganizationInvitations.

        Returns:
            The organization's invitations, sorted by email.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("GET", _INVITATIONS_PATH)
        return [OrganizationInvitation.model_validate(entry) for entry in response.json()]

    def create(self, params: InvitationCreateParams) -> OrganizationInvitation:
        """Invite a new user to join the organization via POST /v2/organization-invitations.

        An invitation email goes out to the invitee. Only one invitation may
        exist per email, and the email may not already have a profile in the
        organization. Mirrors js-sdk createOrganizationInvitation.

        Example:
            invite = endpoint.invitations.create(
                InvitationCreateParams(email="a@b.com", first_name="A", last_name="B", role="member")
            )

        Args:
            params: Details for the invitation.

        Returns:
            The newly-created invitation.

        Raises:
            VerdocsAPIError: The API rejected the request (400 for duplicate
                invitations or existing profiles).
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("POST", _INVITATIONS_PATH, json=_write_body(params))
        return OrganizationInvitation.model_validate(response.json())

    def get(self, email: str, token: str) -> OrganizationInvitation:
        """Get an invitation's details via GET /v2/organization-invitations/{email}/{token}.

        Authenticated by the invite token, not a session; intended for the
        invitee, usually as the first step of accepting. A success means the
        token is still valid, and the response includes the organization's
        summary details for branding the accept screen. Mirrors js-sdk
        getOrganizationInvitation.

        Args:
            email: Email address the invitation was sent to.
            token: The invite token from the invitation email.

        Returns:
            The invitation, including its organization.

        Raises:
            NotFoundError: No invitation matches that email and token.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("GET", f"{_INVITATIONS_PATH}/{email}/{token}")
        return OrganizationInvitation.model_validate(response.json())

    def update(self, email: str, params: InvitationUpdateParams) -> OrganizationInvitation | None:
        """Update a pending invitation via PATCH /v2/organization-invitations/{email}.

        The email itself may not be changed; delete and re-create instead.
        Mirrors js-sdk updateOrganizationInvitation; note the deployed schema
        accepts role only (see InvitationUpdateParams).

        Known server defect: the deployed handler's existence check is
        inverted, so updating an invitation that exists draws a 400 ("An
        invitation already exists for this email") and updating one that does
        not exist "succeeds" with an empty body. The call is unusable until
        the API is fixed; we port it as specified.

        Args:
            email: Email address of the invitation to update.
            params: The fields to change.

        Returns:
            The updated invitation, or None when the wire answers with an
            empty body (see the defect note above).

        Raises:
            VerdocsAPIError: The API rejected the request.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("PATCH", f"{_INVITATIONS_PATH}/{email}", json=_write_body(params))
        payload = response.json() if response.content else None
        if payload is None:
            return None
        return OrganizationInvitation.model_validate(payload)

    def delete(self, email: str) -> None:
        """Delete a pending invitation via DELETE /v2/organization-invitations/{email}.

        No cancellation message is sent; the invitee sees an error if they
        try to join later. The handler also removes any profile rows matching
        that email in the organization. Mirrors js-sdk
        deleteOrganizationInvitation.

        The API answers with a status marker that nothing consumes, so this
        returns None and relies on exceptions for failure.

        Args:
            email: Email address of the invitation to delete.

        Raises:
            VerdocsAPIError: The API returned a non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        self._endpoint._request("DELETE", f"{_INVITATIONS_PATH}/{email}")

    def resend(self, email: str) -> None:
        """Send a reminder to a pending invitee via POST /v2/organization-invitations/resend.

        Declined invitations cannot be resent. Mirrors js-sdk
        resendOrganizationInvitation, which types the response as the
        invitation; the wire answers with a status marker only, so this
        returns None and relies on exceptions for failure.

        Args:
            email: Email address of the invitee to remind.

        Raises:
            NotFoundError: No invitation exists for that email.
            VerdocsAPIError: The API returned another non-2xx status (400 when
                the invitation was declined).
            VerdocsConnectionError: The request never reached the API.
        """
        self._endpoint._request("POST", f"{_INVITATIONS_PATH}/resend", json={"email": email})

    def accept(self, params: InvitationAcceptParams) -> AuthenticateResponse:
        """Accept an invitation via POST /v2/organization-invitations/accept.

        Authenticated by the invite token, not a session. Creates a user
        account for the invitee if needed, creates a profile with the invited
        role, makes it the invitee's current profile, and returns session
        tokens for it; call set_token() with the access token to continue.
        Mirrors js-sdk acceptOrganizationInvitation.

        Example:
            tokens = endpoint.invitations.accept(
                InvitationAcceptParams(
                    email="a@b.com", token="TOKEN", first_name="A", last_name="B", password="secret12"
                )
            )
            endpoint.set_token(tokens.access_token)

        Args:
            params: The invitee's details, invite token, and new password.

        Returns:
            Session tokens for the newly-created profile.

        Raises:
            NotFoundError: No invitation matches that email and token.
            VerdocsAPIError: The API returned another non-2xx status (400 when
                the invitation was declined).
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("POST", f"{_INVITATIONS_PATH}/accept", json=_write_body(params))
        return AuthenticateResponse.model_validate(response.json())

    def decline(self, email: str, token: str) -> None:
        """Decline an invitation via POST /v2/organization-invitations/decline.

        Authenticated by the invite token, not a session. Marks the
        invitation declined, which shows the organization's admins it was
        refused, blocks further invitations to the same email, and stops
        reminders. Mirrors js-sdk declineOrganizationInvitation.

        The API answers with a status marker that nothing consumes, so this
        returns None and relies on exceptions for failure.

        Args:
            email: Email address the invitation was sent to.
            token: The invite token from the invitation email.

        Raises:
            NotFoundError: No invitation matches that email and token.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        self._endpoint._request("POST", f"{_INVITATIONS_PATH}/decline", json={"email": email, "token": token})


class AsyncInvitations:
    """Organization invitation calls for an async endpoint."""

    def __init__(self, endpoint: AsyncVerdocsEndpoint) -> None:
        self._endpoint = endpoint

    async def list(self) -> list[OrganizationInvitation]:
        """Get the pending invitations for the caller's organization via GET /v2/organization-invitations.

        The caller must be an admin or owner. Mirrors js-sdk getOrganizationInvitations.

        Returns:
            The organization's invitations, sorted by email.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", _INVITATIONS_PATH)
        return [OrganizationInvitation.model_validate(entry) for entry in response.json()]

    async def create(self, params: InvitationCreateParams) -> OrganizationInvitation:
        """Invite a new user to join the organization via POST /v2/organization-invitations.

        An invitation email goes out to the invitee. Only one invitation may
        exist per email, and the email may not already have a profile in the
        organization. Mirrors js-sdk createOrganizationInvitation.

        Example:
            invite = await endpoint.invitations.create(
                InvitationCreateParams(email="a@b.com", first_name="A", last_name="B", role="member")
            )

        Args:
            params: Details for the invitation.

        Returns:
            The newly-created invitation.

        Raises:
            VerdocsAPIError: The API rejected the request (400 for duplicate
                invitations or existing profiles).
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("POST", _INVITATIONS_PATH, json=_write_body(params))
        return OrganizationInvitation.model_validate(response.json())

    async def get(self, email: str, token: str) -> OrganizationInvitation:
        """Get an invitation's details via GET /v2/organization-invitations/{email}/{token}.

        Authenticated by the invite token, not a session; intended for the
        invitee, usually as the first step of accepting. A success means the
        token is still valid, and the response includes the organization's
        summary details for branding the accept screen. Mirrors js-sdk
        getOrganizationInvitation.

        Args:
            email: Email address the invitation was sent to.
            token: The invite token from the invitation email.

        Returns:
            The invitation, including its organization.

        Raises:
            NotFoundError: No invitation matches that email and token.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", f"{_INVITATIONS_PATH}/{email}/{token}")
        return OrganizationInvitation.model_validate(response.json())

    async def update(self, email: str, params: InvitationUpdateParams) -> OrganizationInvitation | None:
        """Update a pending invitation via PATCH /v2/organization-invitations/{email}.

        The email itself may not be changed; delete and re-create instead.
        Mirrors js-sdk updateOrganizationInvitation; note the deployed schema
        accepts role only (see InvitationUpdateParams).

        Known server defect: the deployed handler's existence check is
        inverted, so updating an invitation that exists draws a 400 ("An
        invitation already exists for this email") and updating one that does
        not exist "succeeds" with an empty body. The call is unusable until
        the API is fixed; we port it as specified.

        Args:
            email: Email address of the invitation to update.
            params: The fields to change.

        Returns:
            The updated invitation, or None when the wire answers with an
            empty body (see the defect note above).

        Raises:
            VerdocsAPIError: The API rejected the request.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("PATCH", f"{_INVITATIONS_PATH}/{email}", json=_write_body(params))
        payload = response.json() if response.content else None
        if payload is None:
            return None
        return OrganizationInvitation.model_validate(payload)

    async def delete(self, email: str) -> None:
        """Delete a pending invitation via DELETE /v2/organization-invitations/{email}.

        No cancellation message is sent; the invitee sees an error if they
        try to join later. The handler also removes any profile rows matching
        that email in the organization. Mirrors js-sdk
        deleteOrganizationInvitation.

        The API answers with a status marker that nothing consumes, so this
        returns None and relies on exceptions for failure.

        Args:
            email: Email address of the invitation to delete.

        Raises:
            VerdocsAPIError: The API returned a non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        await self._endpoint._request("DELETE", f"{_INVITATIONS_PATH}/{email}")

    async def resend(self, email: str) -> None:
        """Send a reminder to a pending invitee via POST /v2/organization-invitations/resend.

        Declined invitations cannot be resent. Mirrors js-sdk
        resendOrganizationInvitation, which types the response as the
        invitation; the wire answers with a status marker only, so this
        returns None and relies on exceptions for failure.

        Args:
            email: Email address of the invitee to remind.

        Raises:
            NotFoundError: No invitation exists for that email.
            VerdocsAPIError: The API returned another non-2xx status (400 when
                the invitation was declined).
            VerdocsConnectionError: The request never reached the API.
        """
        await self._endpoint._request("POST", f"{_INVITATIONS_PATH}/resend", json={"email": email})

    async def accept(self, params: InvitationAcceptParams) -> AuthenticateResponse:
        """Accept an invitation via POST /v2/organization-invitations/accept.

        Authenticated by the invite token, not a session. Creates a user
        account for the invitee if needed, creates a profile with the invited
        role, makes it the invitee's current profile, and returns session
        tokens for it; call set_token() with the access token to continue.
        Mirrors js-sdk acceptOrganizationInvitation.

        Example:
            tokens = await endpoint.invitations.accept(
                InvitationAcceptParams(
                    email="a@b.com", token="TOKEN", first_name="A", last_name="B", password="secret12"
                )
            )
            endpoint.set_token(tokens.access_token)

        Args:
            params: The invitee's details, invite token, and new password.

        Returns:
            Session tokens for the newly-created profile.

        Raises:
            NotFoundError: No invitation matches that email and token.
            VerdocsAPIError: The API returned another non-2xx status (400 when
                the invitation was declined).
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("POST", f"{_INVITATIONS_PATH}/accept", json=_write_body(params))
        return AuthenticateResponse.model_validate(response.json())

    async def decline(self, email: str, token: str) -> None:
        """Decline an invitation via POST /v2/organization-invitations/decline.

        Authenticated by the invite token, not a session. Marks the
        invitation declined, which shows the organization's admins it was
        refused, blocks further invitations to the same email, and stops
        reminders. Mirrors js-sdk declineOrganizationInvitation.

        The API answers with a status marker that nothing consumes, so this
        returns None and relies on exceptions for failure.

        Args:
            email: Email address the invitation was sent to.
            token: The invite token from the invitation email.

        Raises:
            NotFoundError: No invitation matches that email and token.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        await self._endpoint._request("POST", f"{_INVITATIONS_PATH}/decline", json={"email": email, "token": token})

File: sdks/csharp/src/Verdocs.Sdk/Resources/Invitations.cs
using Verdocs.Models;

namespace Verdocs.Resources;

/// <summary>
/// Organization invitation calls, reached through the endpoint's Invitations property. An
/// invitation is a one-time offer for a person to join an organization; it is keyed by email
/// address and redeemed with the token from the invitation email.
/// </summary>
public sealed class Invitations
{
    private readonly VerdocsEndpoint _endpoint;

    internal Invitations(VerdocsEndpoint endpoint)
    {
        _endpoint = endpoint;
    }

    /// <summary>
    /// Gets the invitations pending for the caller's organization, sorted by email.
    ///
    /// <example>
    /// <code>
    /// var invitations = await endpoint.Invitations.ListAsync();
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The organization's pending invitations.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the session is invalid.</exception>
    /// <sdkOperation>invitation.getOrganizationInvitations</sdkOperation>
    /// <sdkGroup>Invitation</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public async Task<IReadOnlyList<OrganizationInvitation>> ListAsync(CancellationToken cancellationToken = default)
    {
        return await _endpoint.SendAsync<List<OrganizationInvitation>>(
                HttpMethod.Get, "/v2/organization-invitations", null, cancellationToken)
            .ConfigureAwait(false);
    }

    /// <summary>
    /// Invites a person to join the organization and emails them the invitation. The caller
    /// must be an admin. Fails when an invitation or profile already exists for the email.
    ///
    /// <example>
    /// <code>
    /// var invitation = await endpoint.Invitations.CreateAsync(new CreateInvitationRequest
    /// {
    ///     Email = "paige.turner@example.com",
    ///     FirstName = "Paige",
    ///     LastName = "Turner",
    ///     Role = "member",
    /// });
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="request">Details for the invitation.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The new invitation.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because an invitation already exists for the email.</exception>
    /// <sdkOperation>invitation.createOrganizationInvitation</sdkOperation>
    /// <sdkGroup>Invitation</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<OrganizationInvitation> CreateAsync(CreateInvitationRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return _endpoint.SendAsync<OrganizationInvitation>(
            HttpMethod.Post, "/v2/organization-invitations", request, cancellationToken);
    }

    /// <summary>
    /// Deletes a pending invitation. No cancellation message is sent, and the invitee sees an
    /// error if they later try to accept. This also removes any profile rows for that email
    /// in the caller's organization.
    ///
    /// <example>
    /// <code>
    /// await endpoint.Invitations.DeleteAsync("paige.turner@example.com");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="email">The invitee's email address.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>A task that completes when the invitation is deleted.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the caller is not an admin.</exception>
    /// <sdkOperation>invitation.deleteOrganizationInvitation</sdkOperation>
    /// <sdkGroup>Invitation</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task DeleteAsync(string email, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(email);
        return _endpoint.SendVoidAsync(
            HttpMethod.Delete, "/v2/organization-invitations/" + Uri.EscapeDataString(email), null, cancellationToken);
    }

    /// <summary>
    /// Updates a pending invitation's role. The email cannot be changed; delete and re-create
    /// the invitation instead. The js-sdk also sends the names, but the deployed API's strict
    /// schema accepts the role alone. NOTE: the deployed handler's existence check is
    /// inverted, so updating an invitation that exists currently fails with a 400 ("An
    /// invitation already exists for this email") until the API is fixed.
    ///
    /// <example>
    /// <code>
    /// await endpoint.Invitations.UpdateAsync("paige.turner@example.com", "admin");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="email">The invitee's email address.</param>
    /// <param name="role">The role the invitee will hold: "contact", "basic_user", "member", "admin", or "owner".</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>A task that completes when the server accepts the request.</returns>
    /// <exception cref="VerdocsApiException">The call failed; see the note above about the deployed handler.</exception>
    /// <sdkOperation>invitation.updateOrganizationInvitation</sdkOperation>
    /// <sdkGroup>Invitation</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task UpdateAsync(string email, string role, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(email);
        ArgumentException.ThrowIfNullOrEmpty(role);
        return _endpoint.SendVoidAsync(
            HttpMethod.Patch, "/v2/organization-invitations/" + Uri.EscapeDataString(email), new { role }, cancellationToken);
    }

    /// <summary>
    /// Re-sends the invitation email to a pending invitee. Declined invitations cannot be
    /// re-sent. The js-sdk types the response as the invitation, but the server answers with
    /// a status marker only, so nothing is returned here.
    ///
    /// <example>
    /// <code>
    /// await endpoint.Invitations.ResendAsync("paige.turner@example.com");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="email">The invitee's email address.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>A task that completes when the reminder is sent.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the invitation was declined.</exception>
    /// <sdkOperation>invitation.resendOrganizationInvitation</sdkOperation>
    /// <sdkGroup>Invitation</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task ResendAsync(string email, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(email);
        return _endpoint.SendVoidAsync(
            HttpMethod.Post, "/v2/organization-invitations/resend", new { email }, cancellationToken);
    }

    /// <summary>
    /// Gets an invitation's details, typically as the first step of accepting it. The call is
    /// authenticated by the invite token rather than a session, and a successful response
    /// means the token is still valid. Includes the organization's summary details for
    /// branding the acceptance screen.
    ///
    /// <example>
    /// <code>
    /// var invitation = await endpoint.Invitations.GetAsync("paige.turner@example.com", "a1b2c3d4e5f6");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="email">The invitee's email address.</param>
    /// <param name="token">The invite token from the invitation email.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The invitation, with its organization.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the token is invalid.</exception>
    /// <sdkOperation>invitation.getOrganizationInvitation</sdkOperation>
    /// <sdkGroup>Invitation</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<OrganizationInvitation> GetAsync(string email, string token, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(email);
        ArgumentException.ThrowIfNullOrEmpty(token);
        return _endpoint.SendAsync<OrganizationInvitation>(
            HttpMethod.Get,
            "/v2/organization-invitations/" + Uri.EscapeDataString(email) + "/" + Uri.EscapeDataString(token),
            null,
            cancellationToken);
    }

    /// <summary>
    /// Accepts an invitation. Creates a user for the invitee when needed, creates a profile
    /// with the invited role and makes it current, and returns session tokens for it; apply
    /// them with <see cref="VerdocsEndpoint.SetToken"/>.
    ///
    /// <example>
    /// <code>
    /// var auth = await endpoint.Invitations.AcceptAsync(new AcceptOrganizationInvitationRequest
    /// {
    ///     Email = "paige.turner@example.com",
    ///     Token = "a1b2c3d4e5f6",
    ///     FirstName = "Paige",
    ///     LastName = "Turner",
    ///     Password = "correct-horse-battery-staple",
    /// });
    /// endpoint.SetToken(auth.AccessToken);
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="request">The acceptance details.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>Session tokens for the new profile.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the invitation was declined.</exception>
    /// <sdkOperation>invitation.acceptOrganizationInvitation</sdkOperation>
    /// <sdkGroup>Invitation</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<AuthenticateResponse> AcceptAsync(AcceptOrganizationInvitationRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return _endpoint.SendAsync<AuthenticateResponse>(
            HttpMethod.Post, "/v2/organization-invitations/accept", request, cancellationToken);
    }

    /// <summary>
    /// Declines an invitation. The status becomes "declined", which shows the organization's
    /// admins the invite was refused, blocks further invitations to the email, and stops
    /// reminder emails.
    ///
    /// <example>
    /// <code>
    /// await endpoint.Invitations.DeclineAsync("paige.turner@example.com", "a1b2c3d4e5f6");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="email">The invitee's email address.</param>
    /// <param name="token">The invite token from the invitation email.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>A task that completes when the invitation is declined.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the token is invalid.</exception>
    /// <sdkOperation>invitation.declineOrganizationInvitation</sdkOperation>
    /// <sdkGroup>Invitation</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task DeclineAsync(string email, string token, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(email);
        ArgumentException.ThrowIfNullOrEmpty(token);
        return _endpoint.SendVoidAsync(
            HttpMethod.Post, "/v2/organization-invitations/decline", new { email, token }, cancellationToken);
    }
}
